### ISSUE TITLE
Amélioration Fuzzy Search : N-grams contigus pour titres multi-mots

### ISSUE THREAD

**Issue by castorfou** [castorfou/back-office-lmelp#76]
## Problème actuel

Le fuzzy search actuel fragmente les titres en **mots isolés**, ce qui rate les correspondances multi-mots :

**Exemple concret** : `L'invention de Tristan`

```json
{
  "episode_id": "6865f99ba1418e3d7c63d07a",
  "query_title": "L'invention de Tristan",
  "query_author": "Adrien Bosque"
}

// Réponse actuelle (MAUVAISE)
{
  "title_matches": [
    ["📖 https://www.franceinter.fr/...", 42],  // ❌ URL parasite
    ["Tristan", 90],                           // ❌ Fragment isolé
    ["l'aristocratie", 86],                    // ❌ Faux positif
    ["l'humour", 86],                          // ❌ Faux positif
    ["invités", 60]                            // ❌ Bruit
  ]
}

// Réponse attendue (BONNE)
{
  "title_matches": [
    ["📖 L'invention de Tristan", 95],         // ✅ Titre complet
    ["invention de Tristan", 92],              // ✅ Trigram
    ["Tristan", 90]                            // ✅ Fallback
  ]
}
```

**Cause racine** :
```python
# Code actuel (app.py:611-614)
words = [word for word in full_text.split() if len(word) > 3]
search_candidates = quoted_segments + words
# → Fragmentation en mots isolés
```

### 🟢 Étape 1 : N-grams contigus (PRIORITÉ HAUTE)

**Objectif** : Extraire des séquences de N mots consécutifs pour matcher les titres complets

**Implémentation** :

```python
# Fichier : src/back_office_lmelp/app.py:604-620

def extract_ngrams(text: str, n: int) -> list[str]:
    """Extrait des séquences de n mots consécutifs."""
    words = text.split()
    return [' '.join(words[i:i+n]) for i in range(len(words) - n + 1)]

# Après extraction des segments entre guillemets (ligne 608)
quoted_segments = re.findall(r'"([^"]+)"', full_text)

# NOUVEAU : Extraire n-grams de différentes tailles
words = [word for word in full_text.split() if len(word) > 3]
bigrams = extract_ngrams(full_text, 2)    # "L'invention de", "de Tristan"
trigrams = extract_ngrams(full_text, 3)   # "L'invention de Tristan"
quadrigrams = extract_ngrams(full_text, 4) # "L'invention de Tristan Adrien"

# Candidats par priorité : guillemets > 4-grams > 3-grams > 2-grams > mots
search_candidates = (
    quoted_segments +
    [ng for ng in quadrigrams if len(ng) > 10] +  # Filtrer n-grams trop courts
    [ng for ng in trigrams if len(ng) > 8] +
    [ng for ng in bigrams if len(ng) > 6] +
    words
)

# Le reste du code reste identique
all_matches = process.extract(request.query_title, search_candidates, limit=10)
```

**Bénéfices** :
- ✅ Trouve les titres complets en un seul match
- ✅ Pas de dépendance ML (modification mineure)
- ✅ Temps d'implémentation : ~2h
- ✅ Impact immédiat sur les titres multi-mots

**Cas de test à valider** :
```python
# Tests unitaires à ajouter
test_cases = [
    {
        "query": "L'invention de Tristan",
        "episode_text": "Adrien Bosc présente L'invention de Tristan chez Stock",
        "expected_match": "L'invention de Tristan",
        "expected_score": ">= 95"
    },
    {
        "query": "Un autre m'attend ailleurs",
        "episode_text": "Christophe Bigot - Un autre m'attend ailleurs",
        "expected_match": "Un autre m'attend ailleurs",
        "expected_score": ">= 90"
    },
    {
        "query": "Peau d'ourse",
        "episode_text": "Grégory Le Floch Peau d'ourse Seuil",
        "expected_match": "Peau d'ourse",  # ou trigram "Le Floch Peau d'ourse"
        "expected_score": ">= 85"
    }
]
```

**Métriques de succès** :
- **Taux de détection titres complets** : > 80% des titres 3+ mots trouvés en entier
- **Réduction faux positifs** : Moins de fragments isolés dans top 5 résultats
- **Amélioration scores** : Score moyen des vrais titres +10 points minimum

**Fichiers impactés** :
- `src/back_office_lmelp/app.py` (endpoint `/api/fuzzy-search-episode`)
- `tests/test_app.py` (tests unitaires endpoint)
- `docs/dev/biblio-verification-flow.md` (documentation)

**Effort estimé** : 2-3h (implémentation + tests)

---

### 🟡 Étape 2 : Embeddings sémantiques (SI Étape 1 insuffisante)

**Objectif** : Résoudre les homophones et confusions phonétiques Whisper ("tant" vs "temps")

**Prérequis** : Étape 1 testée et résout < 70% des cas difficiles

**Approche** : Modèles de sentence embeddings multilingues

**Modèle recommandé** : `paraphrase-multilingual-MiniLM-L12-v2`
- Taille : ~50MB
- Langue : Français ✅
- Performance : Comprend paraphrases et variations orthographiques

**Implémentation (sketch)** :

```python
# Installation
# pip install sentence-transformers

from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity

# Charger modèle au démarrage de l'app (singleton)
embedding_model = SentenceTransformer('paraphrase-multilingual-MiniLM-L12-v2')

# Dans l'endpoint /api/fuzzy-search-episode
def hybrid_fuzzy_search(query: str, candidates: list[str]) -> list[tuple[str, float]]:
    """Combine fuzzy text matching + semantic similarity."""
    
    # Score 1 : Fuzzy textuel (rapidfuzz)
    fuzzy_matches = process.extract(query, candidates, limit=10)
    fuzzy_scores = {text: score/100 for text, score in fuzzy_matches}
    
    # Score 2 : Similarité sémantique (embeddings)
    query_embedding = embedding_model.encode([query])
    candidates_embeddings = embedding_model.encode(candidates)
    cosine_scores = cosine_similarity(query_embedding, candidates_embeddings)[0]
    semantic_scores = {candidates[i]: cosine_scores[i] for i in range(len(candidates))}
    
    # Score final hybride : 60% textuel + 40% sémantique
    final_scores = {}
    for candidate in candidates:
        fuzzy = fuzzy_scores.get(candidate, 0)
        semantic = semantic_scores.get(candidate, 0)
        final_scores[candidate] = 0.6 * fuzzy + 0.4 * semantic
    
    # Trier par score final
    ranked = sorted(final_scores.items(), key=lambda x: x[1], reverse=True)
    return [(text, score * 100) for text, score in ranked[:10]]
```

**Bénéfices** :
- ✅ Détecte homophones : "tant" vs "temps", "mieu" vs "mieux"
- ✅ Comprend paraphrases : "Un déni français" ≈ "Déni français"
- ✅ Robuste aux accents : "Séphora" ≈ "Sephora"

**Coûts** :
- ❌ Dépendance lourde : `sentence-transformers` (~200MB avec PyTorch)
- ❌ Latence : +50-100ms par requête (vectorisation)
- ❌ Mémoire : +200MB RAM (modèle chargé)

**Métriques de succès** :
- **Homophones détectés** : > 90% des confusions phonétiques Whisper résolues
- **Latence acceptable** : < 200ms total par requête
- **Amélioration vs Étape 1** : +15 points de score moyen sur cas difficiles

**Fichiers impactés** :
- `pyproject.toml` (dépendance `sentence-transformers`)
- `src/back_office_lmelp/app.py` (chargement modèle + endpoint)
- `tests/test_app.py` (tests homophones)

**Effort estimé** : 1 jour (implémentation + optimisation performance + tests)

**⚠️ À ne faire QUE SI** :
- Étape 1 (n-grams) ne résout pas assez de cas (< 70% succès)
- Les homophones Whisper sont un problème récurrent dans les logs
- Performance acceptable avec modèle léger testé

---

### 🔵 Étape 3 : Distance phonétique (CAS EXTRÊMES - Futur)

**Objectif** : Convertir texte en phonèmes pour distance phonétique pure

**Prérequis** : Étape 2 testée et insuffisante sur cas extrêmes

**Approche** : `phonemizer` (français) + distance de Levenshtein phonétique

**Exemple** :
```python
from phonemizer import phonemize

# Conversion texte → phonèmes
text1 = phonemize("tant mieux", language='fr-fr')  # /tɑ̃ mjø/
text2 = phonemize("temps mieu", language='fr-fr')  # /tɑ̃ mjø/

# Distance phonétique
# Levenshtein(/tɑ̃ mjø/, /tɑ̃ mjø/) = 0 → score parfait
```

**Bénéfices** :
- ✅ Homophones parfaits : "tant" = "temps" phonétiquement
- ✅ Accents régionaux : neutralise variations

**Coûts** :
- ❌ Dépendance système : `espeak-ng` (binaire C++)
- ❌ Complexité : installation système + Python
- ❌ Performance : +100ms par requête

**Priorité** : ⭐ BASSE (seulement si Étape 2 insuffisante)

**Effort estimé** : 2 jours (recherche + implémentation + DevOps)

---

### Phase 1 : Quick Win (Semaine 1)
1. ✅ Implémenter Étape 1 (n-grams)
2. ✅ Tests unitaires sur 10 cas réels
3. ✅ Mesurer amélioration vs baseline
4. ✅ Documenter résultats

**Critère de succès** : > 70% des titres multi-mots détectés correctement

### Phase 2 : Décision (Semaine 2)
- **Si succès Étape 1** : ✅ Terminer, documenter, fermer issue
- **Si échec Étape 1** : ➡️ Créer issue séparée pour Étape 2 (embeddings)

### Phase 3 : Optimisation (Si nécessaire)
- Implémenter Étape 2 uniquement si justifiée par données réelles
- Ne PAS implémenter Étape 3 sauf cas extrêmes prouvés

---

### Cas 1 : Titre long multi-mots
```
Query: "L'invention de Tristan"
Episode: "Adrien Bosc L'invention de Tristan Stock"
Expected: Trigram "L'invention de Tristan" trouvé avec score >= 95
```

### Cas 2 : Titre avec apostrophe
```
Query: "Peau d'ourse"
Episode: "Grégory Le Floch Peau d'ourse Seuil"
Expected: Bigram "Peau d'ourse" trouvé avec score >= 90
```

### Cas 3 : Homonyme Whisper (nécessite Étape 2)
```
Query: "Tant mieux"
Episode (Whisper): "Amélie Nothomb Tant mieu Albin Michel"
Expected (Étape 1): Pas de match
Expected (Étape 2): Semantic match "Tant mieu" ≈ "Tant mieux" score >= 85
```

### Cas 4 : Inversion nom (HORS SCOPE - nécessite autre approche)
```
Query: "Grégory Lefloc"
Episode: "Grégory Le Floch"
Expected: N-grams ne peuvent pas résoudre (segmentation différente)
Note: Nécessiterait mapping canonique auteurs (issue séparée)
```

---

## Références

- Issue #75 : Amélioration Phase 0 (double appel Babelio)
- Issue #74 : Filtrage suggestions invalides (URLs, fragments)
- Documentation : biblio-verification-flow.md Phase 1

---

**Labels** : `enhancement`, `fuzzy-search`, `performance`
**Priorité** : Haute (Étape 1), Moyenne (Étape 2), Basse (Étape 3)
**Effort** : 2-3h (Étape 1), 1 jour (Étape 2), 2 jours (Étape 3)
**Impact estimé** : +40-60% précision fuzzy search (Étape 1 seule)
